### ISSUE TITLE
Cannot perform state.highstate/saltutil.sync_all on proxy minion

### ISSUE THREAD

**Issue by toddtomkinson** [saltstack/salt#26786]
I am trying to create states for a proxy minion.  The proxy minion itself comes up fine and I can tell that communication is working between the master, the proxy minion, and my remote resource (a ping that hits the remote resource is successful).  I cannot, however, get states to work because whenever I run a highstate or even a saltutils.sync_* I get the following error on the minion:
```python-traceback
    [WARNING ] TypeError encountered executing saltutil.sync_all: can't serialize <salt.loader.LazyLoader object at 0x166f350>. See debug log for more info.
    Traceback (most recent call last):
      File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1173, in _thread_return
        return_data = func(*args, **kwargs)
      File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 421, in sync_all
        ret['beacons'] = sync_beacons(saltenv, False)
      File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 263, in sync_beacons
        ret = _sync('beacons', saltenv)
      File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 88, in _sync
        saltenv = _get_top_file_envs()
      File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 72, in _get_top_file_envs
        envs = list(st_.top_matches(top).keys()) or 'base'
      File "/usr/lib/python2.7/site-packages/salt/state.py", line 2472, in top_matches
        ext_matches = self.client.ext_nodes()
      File "/usr/lib/python2.7/site-packages/salt/fileclient.py", line 1233, in ext_nodes
        return self.channel.send(load)
      File "/usr/lib/python2.7/site-packages/salt/transport/__init__.py", line 314, in send
        return self._crypted_transfer(load, tries, timeout)
      File "/usr/lib/python2.7/site-packages/salt/transport/__init__.py", line 302, in _crypted_transfer
        return _do_transfer()
      File "/usr/lib/python2.7/site-packages/salt/transport/__init__.py", line 291, in _do_transfer
        self.auth.crypticle.dumps(load),
      File "/usr/lib/python2.7/site-packages/salt/crypt.py", line 835, in dumps
        return self.encrypt(self.PICKLE_PAD + self.serial.dumps(obj))
      File "/usr/lib/python2.7/site-packages/salt/payload.py", line 121, in dumps
        return msgpack.dumps(msg)
      File "/usr/lib64/python2.7/site-packages/msgpack/__init__.py", line 47, in packb
        return Packer(**kwargs).pack(o)
      File "msgpack/_packer.pyx", line 223, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:223)
      File "msgpack/_packer.pyx", line 225, in msgpack._packer.Packer.pack (msgpack/_packer.cpp:225)
      File "msgpack/_packer.pyx", line 184, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:184)
      File "msgpack/_packer.pyx", line 184, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:184)
      File "msgpack/_packer.pyx", line 220, in msgpack._packer.Packer._pack (msgpack/_packer.cpp:220)
    TypeError: can't serialize <salt.loader.LazyLoader object at 0x166f350>
```
I'm using salt 2015.5.5 and have narrowed this down to adding the proxy module LazyLoader object to the minion's opts dictionary happening here:

https://github.com/saltstack/salt/blob/v2015.5.5/salt/minion.py#L2948

I was able to perform a highstate by changing the code in the class referenced above to just use the self.proxymodule handle to reference the LazyLoader object, but then I realized that the test.ping function no longer worked correctly because it was looking for the 'proxymodule' LazyLoader in the __opts__ dictionary (https://github.com/saltstack/salt/blob/develop/salt/modules/test.py#L113).  The other modules in salt.proxy also rely on the 'proxymodule' object being in the __opts__dictionary.

I haven't dug any further than that and wanted to see if someone else could talk about what is going on here.  This is reproducible with the standard salt proxy minion rest example (running saltutil.sync_all or state.highstate).

**Comment by jfindlay** [comment]
@toddtomkinson, thanks for the report.  @cro or @cachedout may have more to say about the issues you're seeing.

**Comment by cro** [comment]
Hi @toddtomkinson,

Thanks for the deep code-dive here, this helps a lot in sorting this out.

I actually get a different stacktrace than you do when doing a sync_all:

```
rest_sample-localhost:
    The minion function caused an exception: Traceback (most recent call last):
      File "/vagrant/salt/salt/minion.py", line 1173, in _thread_return
        return_data = func(*args, **kwargs)
      File "/vagrant/salt/salt/modules/saltutil.py", line 421, in sync_all
        ret['beacons'] = sync_beacons(saltenv, False)
      File "/vagrant/salt/salt/modules/saltutil.py", line 263, in sync_beacons
        ret = _sync('beacons', saltenv)
      File "/vagrant/salt/salt/modules/saltutil.py", line 108, in _sync
        __salt__['cp.cache_dir'](
      File "/vagrant/salt/salt/utils/lazy.py", line 87, in __getitem__
        return self._dict[key]
    KeyError: 'cp.cache_dir'
```

I need to dig further into this and also do some testing to see if the same problem is in the 2015.8 branch.

**Comment by toddtomkinson** [comment]
@cro, I ran into that one, too, and it just requires the following line be added to the cp module:

    __proxyenabled__ = ['*']

I got the same exception you did after I removed the 'proxymodule' from the opts dictionary.

I'm troubled that you didn't get the same exception I did, though.  You were testing on 2015.5.5?  I'll go back through my environment and check things out.

**Comment by toddtomkinson** [comment]
@cro, I was able to take another look at this tonight.  I was not able to reliably reproduce using saltutil.sync_all, but was with state.highstate.  I created a Vagrantfile that sets up an environment to reproduce here:

https://github.com/toddtomkinson/salt-contrib/blob/issue-26786/proxyminion_rest_example/Vagrantfile

You should be able to add the Vagrantfile in that directory to your local salt-contrib source, run 'vagrant up', then see the error in the provisioner output.  From there you can 'vagrant ssh' to log into the box and debug some more.

**Comment by toddtomkinson** [comment]
@cro, I ran the same tests with the 2015.8 branch.  It looks like things have changed significantly in the way proxy minions work.  I like the idea of having a separate salt-proxy call, but it looks like that isn't included as a script in the setup.py yet so it isn't being installed with the bootstrap.

When I copied the salt-proxy script into place then tweaked the pillar data I started getting this error, which is similar to the one above (looks like the same root cause):

    [root@salt ~]# salt \* state.highstate
    rest_sample:
        Passed invalid arguments to state.highstate: can't serialize <salt.loader.LazyLoader object at 0x17c3550>
    
            Retrieve the state data from the salt master for this minion and execute it
    ...

And when I try a saltutil.sync_all:

    [root@salt ~]# salt \* saltutil.sync_all
    rest_sample:
        Passed invalid arguments to saltutil.sync_all: can't serialize <salt.loader.LazyLoader object at 0x28e2110>
    
            Sync down all of the dynamic modules from the file server for a specific
            environment
    ...

Can you reproduce it?  Any other direction I can take?

Thanks!

**Comment by cro** [comment]
@toddtomkinson The issue with setup.py is fixed now, I just merged a PR to take care of it.  I'll see if I can get some time to figure out what's causing the 'invalid arguments' error.

**Comment by toddtomkinson** [comment]
@cro, I was able to get this working with the following patch:

    diff --git a/salt/fileclient.py b/salt/fileclient.py
    index 5805acb..80296cf 100644
    --- a/salt/fileclient.py
    +++ b/salt/fileclient.py
    @@ -1242,9 +1242,11 @@ class RemoteClient(Client):
             Return the metadata derived from the external nodes system on the
             master.
             '''
    +        _opts = dict(self.opts)
    +        _opts.pop('proxymodule')
             load = {'cmd': '_ext_nodes',
                     'id': self.opts['id'],
    -                'opts': self.opts}
    +                'opts': _opts}
             if self.auth:
                 load['tok'] = self.auth.gen_token('salt')
             return self.channel.send(load)
    diff --git a/salt/modules/cp.py b/salt/modules/cp.py
    index f8ebf36..4d6e9be 100644
    --- a/salt/modules/cp.py
    +++ b/salt/modules/cp.py
    @@ -22,6 +22,7 @@ from salt.exceptions import CommandExecutionError
     import salt.ext.six as six

     log = logging.getLogger(__name__)
    +__proxyenabled__ = ['*']


     def _auth():

I don't think the changes in the fileclient module are what should be done, but I don't know enough about what is done on the other side to know exactly which opts values are required (it looked like at least grains--https://github.com/saltstack/salt/blob/e252473b7aeb9947e40093c40b1d5647b22aca0b/salt/daemons/masterapi.py#L512).  I thought I'd pass on what I found to see if it makes things easier to troubleshoot.

**Comment by hishailesh77** [comment]
I am facing the same issue in latest release i.e. '2015.8' . Is there any workaround for the same ?.

**Comment by toddtomkinson** [comment]
@hishailesh77, we are currently using a patch similar to the one I included above to get this working until it is fixed.  The patch is only needed on the hosts running the proxy minion.

**Comment by cro** [comment]
The "real" fix looks like it is finally passing tests: #27731

We'll do some manual tests on it as well and get it merged, hopefully today.

**Comment by rallytime** [comment]
@toddtomkinson or @cro is this bug fixed with #27731 or does more work need to be done here?

**Comment by cro** [comment]
It should be fixed.


On Tue, Oct 20, 2015 at 10:10 AM, Nicole Thomas <notifications@github.com>
wrote:

> @toddtomkinson <https://github.com/toddtomkinson> or @cro
> <https://github.com/cro> is this bug fixed with #27731
> <https://github.com/saltstack/salt/pull/27731> or does more work need to
> be done here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/issues/26786#issuecomment-149618392>.
>



-- 
*--cro*
C. R. Oldham, Platform Engineer, SaltStack
cr@saltstack.com / 801-564-4673
